### PR TITLE
feat(ESSNTL-3533): Switch tag filtering to use OR logic instead of AND

### PR DESCRIPTION
**Pull request by kruai**
# Overview

This PR is being created to address [ESSNTL-3533](https://issues.redhat.com/browse/ESSNTL-3533).
The requirements for tag-filtering have changed. Originally, if multiple tags were filtered on, only hosts that had _all_ of the provided tags would be returned. Instead, they want it so that when multiple tags are provided to the filter, all hosts with _any_ of the provided filters are returned.

I updated the logic in HBI, and updated the expected results for some unit tests to match the logic that works in xjoin queries. I've verified the logical structure of the updated xjoin queries by running them locally in the playground.

Note that in some of the unit tests, the expected results have an OR inside an AND, which can look strange at first. But to help clarify:
```
"AND": ({         // This is the top-level AND that tells elasticsearch that ALL of the contained criteria must be met.
                  // The tags criteria must be met, AND maybe an FQDN needs to match, AND a display_name, etc.

  "OR": (         // This is just one of the criteria. Only ONE of the tags needs to match, but then all other criteria
                  // within the AND must be satisfied.
    { <tag1> },
    { <tag2> },
  )
},),
```

Also, since I had to make these changes in multiple areas, I decided to refactor chunks of the filtering and standardize our graphql queries. Now, the filter criteria that are added are always added in the same order, and if we need to change our filtering, we only need to do that in one place.

## PR Checklist

- [x] Keep PR title short, ideally under 72 characters
- [ ] Descriptive comments provided in complex code blocks
- [x] Tests: validate optimal/expected output
- [x] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Uses [type hinting](https://docs.python.org/3/library/typing.html), if convenient
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: tests/test_xjoin.py [hunk 1]
```diff
@@ -1720,7 +1771,7 @@ def test_xjoin_search_using_hostfilters_tags(
     tag_filters = tuple({"tag": item} for item in tags)
 
     graphql_query_empty_response.assert_called_once_with(
-        HOST_IDS_QUERY, {"filter": tag_filters, "limit": mocker.ANY, "offset": 0}, mocker.ANY
+        HOST_IDS_QUERY, {"filter": ({"OR": tag_filters},), "limit": mocker.ANY, "offset": 0}, mocker.ANY
```
Comment (CodeHeeler) [inline tests/test_xjoin.py:1774]: What about the tag_filters in the test for the `test_system_profile_sap_system_endpoint_tags` on line 1212 of the original, 1263 of your PR?

**Review by CodeHeeler — DISMISSED** [review]
Thanks for this PR! I see what you did with the tag_filter and tag instances within the tests to get those passing.

Had a couple questions on it, otherwise looks good.

File: api/filtering/filtering.py [hunk 2]
```diff
@@ -374,7 +374,8 @@ def query_filters(
         query_filters = ()
 
     if tags:
-        query_filters += build_tag_query_dict_tuple(tags)
+        tag_filters = tuple(build_tag_query_dict_tuple(tags))
```
[the remark this answers is not in the document]
Comment (kruai) [reply api/filtering/filtering.py:377]: Good question! The answer is negligent copypasta - it's redundant and I'll remove it :)

File: tests/test_xjoin.py (hunk 1 above)
Comment (kruai) [reply tests/test_xjoin.py:1774]: Ah, good point - this change needs to also work for the SAP system queries. I'll make some additional changes

Comment (kruai) [reply tests/test_xjoin.py:1774]: OK, I've updated it for SAP queries as well. I might try to refactor some of that file's filter code while I'm at it

Comment (CodeHeeler) [reply tests/test_xjoin.py:1791]: Cool, thanks!

**Comment by CodeHeeler** [comment]
Builds are failing as unstable.  Also, iqe tests likely won't pass until updated to accommodate this work, see related tasks ESSNTL-3997, ESSNTL-3998. @fstavela @computercamplove fyi ^

**Comment by kruai** [comment]
> Builds are failing as unstable. Also, iqe tests likely won't pass until updated to accommodate this work, see related tasks ESSNTL-3997, ESSNTL-3998. @fstavela @computercamplove fyi ^

I expect that for the "tags" IQE failures I see in the build, but I don't expect the error 500s I'm seeing for the `test_filter_tags` IQE tests. I can't retrieve the run results from ibutsu for some reason; do y'all know what the API requests that failed were?

**Comment by fstavela** [comment]
> > Builds are failing as unstable. Also, iqe tests likely won't pass until updated to accommodate this work, see related tasks ESSNTL-3997, ESSNTL-3998. @fstavela @computercamplove fyi ^
> 
> I expect that for the "tags" IQE failures I see in the build, but I don't expect the error 500s I'm seeing for the `test_filter_tags` IQE tests. I can't retrieve the run results from ibutsu for some reason; do y'all know what the API requests that failed were?

Yes, that's right, the PR checks won't pass until the IQE tests are updated. There is quite a lot of work on the squads right now, but I hope to get it done by the end of the week.

The request erroring with 500 doesn't seem right. Here are the details:
```
method = 'GET'
url = 'http://host-inventory-service.ephemeral-b8quv5.svc:8000/api/inventory/v1/tags'
query_params = [('search', 'hbi-filter-test')]
```
So `/tags?search=hbi-filter-test`

**Review by thearifismail — APPROVED** [review]
LGTM

**Comment by kruai** [comment]
Something in the refactor was breaking, and it's difficult to pinpoint what at the moment. I've reverted the refactor, and will open a separate PR for it after this one goes in.

**Review by CodeHeeler — APPROVED** [review]
Splitting out the refactor makes sense.  The rest still looks good.

**Pull request merged**